### PR TITLE
Add architecture overview

### PR DESCRIPTION
**Pull request by willm30**
This PR adds an introduction to the CDapp's architecture.

I feel like it could be improved and so have opened the following issues:

1. https://github.com/JoinColony/ColonyFrontEndLivingStandard/issues/8
2. https://github.com/JoinColony/ColonyFrontEndLivingStandard/issues/9
3. https://github.com/JoinColony/ColonyFrontEndLivingStandard/issues/10

Please assign yourself if you'd like to improve these docs! 🙂 (and add more you think may be missing)

As for this PR specifically, please commit proposed changes directly to this branch.

### CODE REVIEW

File: architecture/index.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Architecture overiew
+
+## How data flows through the CDapp
+
+In principle, we want to handle all interactions with the chain outside of the client. We do this using the block ingestor and lambda functions.
+
+### The Block Ingestor
```
Comment (rdig) [inline architecture/index.md:7]: It needs a section on "why' we chose this approach.

TL;DR; Because we want to have a centralized (ie: faster) API, but we also want the data to be read-only, meaning the chain is the source of truth

File: architecture/index.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+A very common pattern at Colony is taking some data from the user in the ui (usually via a form), and then using that data to call some method in the `colony-js` library, which in turn calls the respective method on the colony network contracts (written in solidity). We coordinate all of this via `redux` actions.
+
+At a glance:
+
+1. Use the ActionForm or ActionButton components to dispatch an action to the store
+2. That action will get picked up by the `redux-saga` middleware, and if there's a listener attached, will run the corresponding saga
+3. Inside the saga, we interact with our ColonyNetwork contracts using `colony-js`. We also dispatch actions to the store every time the state of a transaction changes, so the `GasStation` can display this information to the user.
```
Comment (rdig) [inline architecture/index.md:15]: This needs to specifically mention Chain Transactions, as they are the import part in this equation. "how" we get them sent are implementation minutiae

File: architecture/index.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+4. If the side effect triggers an on-chain event, we listen for that event in a separate service called the [block ingestor](https://github.com/JoinColony/block-ingestor). Generally, we handle all database mutations inside the block-ingestor.
+5. The ui will then implement some kind of "optimstic ui" with the intended changes, or poll the database until the block-ingestor's changes show up.
+
+It looks something like:
+
+![CDapp data flow](https://user-images.githubusercontent.com/1193222/213923691-9f7f6512-861d-4e0a-97d3-4997f7e0d0cf.png)
+
+### Lambda functions
```
Comment (rdig) [inline architecture/index.md:23]: Again, a section as to "why"

TL;DR; We're using them to offload chain calls onto the "server" as to speed up client execution and data updates

**Review by rdig — CHANGES_REQUESTED** [review]
I may be add more entries here with time, but it's good enough for now.

I've left notes that need to be added.

File: architecture/index.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+Sometimes you need on-chain data that either isn't saved in the db, or doesn't itself trigger an event to be picked up in the block ingestor.
```
Comment (ArmandoGraterol) [inline architecture/index.md:25]: I think a fair question that could come out of reading this is why is some data stored in the DB and some are not. An example could be, why are we not storing reputation and instead choosing to get it via `colonyJS` using lambda functions directly without going to the DB in any part of this process. 

Would this be something covered by the DB section of this overview? What do you guys think?

**Review by ArmandoGraterol — COMMENTED** [review]
Looking good and with the sections that you mentioned in the PR's description, the architecture should not be too complicated to understand :)

I left a question regarding the lamda function section

File: architecture/index.md (hunk 4 above)
Comment (willm30) [reply architecture/index.md:25]: @ArmandoGraterol Good question. @rdig?

**Comment by willm30** [comment]
@rdig 
> it's good enough for now.

What should I do to get this in the green from your perspective? Create issues for the improvements you're suggesting? 

I strongly want to avoid the temptation to have me write everything (just because I got it started), as I think this will only work long term if we all get into the habit of contributing.

File: architecture/index.md (hunk 4 above)
Comment (rdig) [reply architecture/index.md:25]: Good point however that can only be properly explained on a case by case basis. (Ie: reputation, since we can't cache it properly in the db)

But for this to work, it would need us to list all existing lambdas, and maintain that list as time goes by, which I don't think is the point of this doc.

We might write some generic line here like: "we choose lambdas over db in cases where the data can't be cached or when it needs to be fetched directly from the chain at the time of the request"

**Comment by willm30** [comment]
Have updated to include diagrams and feedback mentioned here. May still need improving, e.g. with references to things like the Auth server and the Beta Invite system that I'm not as familiar with

**Review by jakubcolony — APPROVED** [review]
Great work!
